grug/malachite/src/actors/host/state.rs

```diff
@@ -0,0 +1,218 @@
+use {
+    crate::{
+        ActorResult, HostConfig, MempoolConfig, RawTx,
+        app::HostAppRef,
+        context::Context,
+        ctx,
+        types::{Block, BlockHash, DecidedBlock},
+    },
+    grug::{Batch, IndexedMap, Map, Storage, UniqueIndex},
+    grug_app::{AppResult, ConsensusStorage},
+    malachitebft_app::{consensus::Role, streaming::StreamId, types::ProposedValue},
+    malachitebft_core_types::Round,
+    malachitebft_engine::consensus::ConsensusRef,
+    std::time::{Duration, Instant},
+};
+
+pub type HeightKey = u64;
+pub type RoundKey = i64;
+
+#[derive(Clone)]
+pub struct State {
+    pub height: ctx!(Height),
+    pub proposer: Option<ctx!(Address)>,
+    pub role: Role,
+    pub round: Round,
+    pub config: HostConfig,
+    pub mempool_config: MempoolConfig,
+    pub db_storage: Box<dyn ConsensusStorage>,
+    consensus: Option<ConsensusRef<Context>>,
+    started_round: Instant,
+    app: HostAppRef,
+    pending_commit_block_hash: Option<BlockHash>,
+}
+
+impl State {
+    pub fn new<S: ConsensusStorage + 'static>(
+        storage: S,
+        config: HostConfig,
+        mempool_config: MempoolConfig,
+        app: HostAppRef,
+    ) -> Self {
+        Self {
+            db_storage: Box::new(storage),
+            consensus: None,
+            height: <ctx!(Height)>::new(1),
+            round: Round::Nil,
+            proposer: None,
+            role: Role::None,
+            config,
+            mempool_config,
+            started_round: Instant::now(),
+            app,
+            pending_commit_block_hash: None,
+        }
+    }
+
+    pub fn set_consensus(&mut self, consensus: ConsensusRef<Context>) {
+        self.consensus = Some(consensus);
+    }
+
+    pub fn consensus(&self) -> ActorResult<ConsensusRef<Context>> {
+        self.consensus
+            .clone()
+            .ok_or(anyhow::anyhow!("Consensus not set").into())
+    }
+
+    pub fn stream_id(&self) -> StreamId {
+        let mut bytes = Vec::with_capacity(size_of::<u64>() + size_of::<u32>());
+        bytes.extend_from_slice(&self.height.to_be_bytes());
+        // TODO: can this panic?
+        bytes.extend_from_slice(&self.round.as_u32().unwrap().to_be_bytes());
```

Comment:
Not in practice given the message flow. If not defined (`Round::Nil`) it will panic, otherwise the round is `u32`.
But maybe nice to anyhow instead of unwrap.